hail/python/hail/vds/combiner.py

```diff
@@ -0,0 +1,216 @@
+                                                              hl.missing(e.PL.dtype)),
+                                                   hl.if_else(alleles_len > 1,
+                                                              e.PL,
+                                                              hl.missing(e.PL.dtype)))
+                handled_fields['RGQ'] = hl.if_else(
+                    has_non_ref,
+                    hl.if_else(e.GT.is_haploid(),
+                               e.PL[alleles_len - 1],
+                               e.PL[hl.call(0, alleles_len - 1).unphased_diploid_gt_index()]),
+                    hl.missing(e.PL.dtype.element_type))
+
+            handled_fields['gvcf_info'] = (hl.case()
+                                           .when(hl.is_missing(row.info.END),
+                                                 hl.struct(**(
+                                                     parse_as_fields(
+                                                         row.info.select(*info_to_keep),
+                                                         has_non_ref)
+                                                 )))
+                                           .or_missing())
+
+            pass_through_fields = {k: v for k, v in e.items() if k not in handled_names}
+            return hl.struct(**handled_fields, **pass_through_fields)
+
+        f = hl.experimental.define_function(
+            lambda row: hl.rbind(
+                hl.len(row.alleles), '<NON_REF>' == row.alleles[-1],
+                lambda alleles_len, has_non_ref: hl.struct(
+                    locus=row.locus,
+                    alleles=hl.if_else(has_non_ref, row.alleles[:-1], row.alleles),
+                    rsid=row.rsid,
+                    __entries=row.__entries.map(
+                        lambda e: make_entry_struct(e, alleles_len, has_non_ref, row)))),
+            mt.row.dtype)
+        _transform_variant_function_map[mt.row.dtype] = f
+    transform_row = _transform_variant_function_map[mt.row.dtype]
+    return unlocalize(Table(TableMapRows(mt._tir, Apply(transform_row._name, transform_row._ret_type, TopLevelReference('row')))))
+
+
+def defined_entry_fields(mt: MatrixTable, sample=None) -> Set[str]:
+    if sample is not None:
+        mt = mt.head(sample)
+    used = mt.aggregate_entries(hl.struct(**{
+        k: hl.agg.any(hl.is_defined(v)) for k, v in mt.entry.items()
+    }))
+    return set(k for k in mt.entry if used[k])
+
+
+def make_reference_matrix_table(mt: MatrixTable, entry_to_keep) -> MatrixTable:
+    mt = mt.filter_rows(hl.is_defined(mt.info.END))
+
+    if not entry_to_keep:
+        # set difference here, since all GTs are 0/0 and for a homref call,
+        # LAD and LPL are uninteresting and trivially recreatable assuming
+        # DP is defined (which it should be!).
+        entry_to_keep = defined_entry_fields(mt, sample=10_000) - {'AD', 'GT', 'PL'}
```

Comment:
I think we shouldn't make this transformation by default, especially unchecked -- we should strive to leave the data model basically untouched with respect to the input gvcfs. This might drop data (it definitely would for the 1kg gvcfs we use in tests).

Happy to drop GT (we should probably insert a runtime check that it's 0/0 if the END field is defined), but I think we should let the others go through the default handling (leave them in ref data if they're defined at ref blocks)

Reply:
I did an experiment, for every 1kg gvcf, after the initial transformation, every reference LAD or LPL was either missing or `LAD == [DP]` and `LPL == [0]`

Reply:
I realize that *our* data generally has this property, but I still object to building in this assumption. It means we're tailoring the combiner specifically to GATK output (in a certain version range likely), and leads to lossy transformations of certain inputs.

Reply:
I do still feel strongly about this point

Reply:
I think ready to go once we fix this

Reply:
We drop PL for reference calls in the current VCF combiner. I've removed the removal of AD in favor of proper handling with translation to LAD (if it's there). For PL, I could have an LPL array, which would change the behavior from the current combiner.